warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java

```diff
@@ -2248,6 +2271,19 @@ else if (IvaratorRequiredVisitor.isIvaratorRequired(queryTree) && !config.canHan
         return new Tuple2<>(ranges, needsFullTable);
     }
     
+    private void attemptToRecordPlan(ShardQueryConfiguration config, JexlNode queryTree) {
+        if (null == this.plannedScript) {
+            try {
+                this.plannedScript = JexlStringBuildingVisitor.buildQuery(queryTree);
+            } catch (Exception e) {
+                if (log.isTraceEnabled()) {
+                    log.trace("Unable to set plannedScript.");
+                }
+            }
+        }
+        config.setQueryString(this.plannedScript);
```

Comment:
Shouldn't we update the plan with each call?  This is only going to update the planned script if it is null, which it won't be after the first time it's set, right?  Am I missing something?

Reply:
Personally, I would simplify this and just call config.setQueryString(JexlStringBuildingVisitor.buildQuery(queryTree)) after each visitor visits the tree.  You might also want to add some kind of indicator which makes it clear that this is the partial query plan, and not the final query plan.

Reply:
Looking some more I see that you only update the query plan on a known failure.  Instead of doing that I would just update it every step of the way, and know that you will be covered if an exception is thrown at any point in the planner.

Reply:
Sounds reasonable

Reply:
I considered updating it on each change to the queryTree but thought it might be too expensive.  If you both think that's acceptable, I'll change it to work that way.

Reply:
printing the tree is orders of magnitude faster than looking things up in the index.  i doubt we will notice.  however, perhaps it is worth creating a ticket to have each visitor return a boolean saying whether or not the tree was modified by the visitor.  if it wasn't modified, then you could skip updating the plan string.  that's a bigger task though, obviously.  worth creating a ticket?